src/librustdoc/passes/unindent_comments.rs

```diff
@@ -35,52 +34,69 @@ impl clean::Attributes {
-
-        let min_indent = if ignore_previous_indents { usize::MAX } else { min_indent };
-
-        if saw_first_line {
-            saw_second_line = true;
-        }
+    // `add` is used in case the most common sugared doc syntax is used ("/// "). The other
+    // fragments kind's lines are never starting with a whitespace unless they are using some
+    // markdown formatting requiring it. Therefore, if the doc block have a mix between the two,
+    // we need to take into account the fact that the minimum indent minus one (to take this
+    // whitespace into account).
+    //
+    // For example:
+    //
+    // /// hello!
+    // #[doc = "another"]
+    //
+    // In this case, you want "hello! another" and not "hello!  another".
+    let add = if !docs.windows(2).all(|arr| arr[0].kind == arr[1].kind)
+        && docs.iter().any(|d| d.kind == DocFragmentKind::SugaredDoc)
+    {
+        // In case we have a mix of sugared doc comments and "raw" ones, we want the sugared one to
+        // "decide" how much the minimum indent will be.
+        1
+    } else {
+        0
+    };
 
-        if line.chars().all(|c| c.is_whitespace()) {
-            min_indent
-        } else {
-            saw_first_line = true;
-            let mut whitespace = 0;
-            line.chars().all(|char| {
-                // Compare against either space or tab, ignoring whether they
-                // are mixed or not
-                if char == ' ' || char == '\t' {
-                    whitespace += 1;
-                    true
+    // `min_indent` is used to know how much whitespaces from the start of each lines must be
+    // removed. Example:
+    //
+    // ///     hello!
+    // #[doc = "another"]
+    //
+    // In here, the `min_indent` is 1 (because non-sugared fragment are always counted with minimum
+    // 1 whitespace), meaning that "hello!" will be considered a codeblock because it starts with 4
+    // (5 - 1) whitespaces.
+    let min_indent = match docs
+        .iter()
+        .map(|fragment| {
+            fragment.doc.lines().fold(usize::MAX, |min_indent, line| {
+                if line.chars().all(|c| c.is_whitespace()) {
+                    min_indent
                 } else {
-                    false
+                    // Compare against either space or tab, ignoring whether they are
+                    // mixed or not.
+                    let whitespace = line.chars().take_while(|c| *c == ' ' || *c == '\t').count();
+                    cmp::min(min_indent, whitespace)
+                        + if fragment.kind == DocFragmentKind::SugaredDoc { 0 } else { add }
                 }
-            });
-            cmp::min(min_indent, whitespace)
-        }
-    });
+            })
+        })
+        .min()
+    {
+        Some(x) => x,
+        None => return,
+    };
+
+    for fragment in docs {
+        let lines: Vec<_> = fragment.doc.lines().collect();
```

Comment:
nit: this seems a little inefficient, I think you only need the collect for the `is_empty()` call? Instead you could check `fragment.doc.trim().is_empty()` and use `fragment.doc.lines()` for the loop.

Reply:
We want to keep the empty lines, they're important in the markdown. :)

Reply:
Right, I'm saying you can keep the empty lines without having to collect into a vec, like this:
```suggestion
if fragment.doc.lines().count() != 0 {
    fragment.doc = fragment.doc.lines().map(...);
```